### PR TITLE
Add CSV Importer

### PR DESCRIPTION
**Pull request by ruslandoga**
### Changes

This PR  implements exports from `events_v2` and `sessions_v2` and imports into `imported_*` tables.
 
- [ ] check export queries, are they correct, compare with CSV report queries (name=pageview filtering, etc.), ensure that after exporting website.com and then importing it as imported.website.com, the dashboards for website.com and imported.website.com show the same statistics
- [ ] tests tests tests
- [ ] adapt to #3724 
- [ ] `import_id` and [staging tables](https://github.com/plausible/analytics/pull/3692#discussion_r1458239447)
- [ ] ~~[zstd? Native?](https://github.com/plausible/analytics/pull/3692#discussion_r1464180510)~~
- [ ] ~~flash message with success or error on import~~

### Tests
- [ ] Automated tests have been added

### Changelog
- [ ] Entry has been added to changelog

### Documentation
- [ ] [Docs](https://github.com/plausible/docs) have been updated

### Dark mode
- [ ] The UI has been tested both in dark and light mode

### CODE REVIEW

File: lib/plausible/data_migration/numeric_ids.ex [hunk 1]
```diff
@@ -85,7 +85,7 @@ defmodule Plausible.DataMigration.NumericIDs do
       end
 
     drop_v2_extra_opts = fn table ->
-      case @repo.query("select count(*) from {$0:Identifier}", [table]) do
+      case @repo.query("select count(*) from {table:Identifier}", %{"table" => table}) do
```
Comment (ruslandoga) [inline lib/plausible/data_migration/numeric_ids.ex:88]: I guess I was a bit too radical with these "named" params changes. I'll revert them back.

File: lib/plausible_web/controllers/site_controller.ex [hunk 2]
```diff
@@ -862,4 +862,77 @@ defmodule PlausibleWeb.SiteController do
...
+        |> Plausible.Site.start_import(~D[0001-01-01], ~D[2030-01-01], "CSV imports")
+        |> Plausible.Repo.update!()
+
+      site
+      |> Plausible.Site.import_success()
+      |> Plausible.Repo.update!()
+    end)
+
+    conn
+    |> put_flash(:info, "IMPORT SUCCESS")
```
Comment (ruslandoga) [inline lib/plausible_web/controllers/site_controller.ex:919]: - [ ] seems like there are no flashes on that page

**Comment by ruurtjan** [comment]
Thanks for your work on this!

I'd be happy to donate $100 when this lands in both the cloud version and the Docker image. DM me then :)

File: lib/plausible_web/controllers/site_controller.ex [hunk 3]
```diff
@@ -862,4 +862,87 @@ defmodule PlausibleWeb.SiteController do
...
+      conn
+      |> put_resp_content_type("application/zip")
+      |> put_resp_header("content-disposition", ~s|attachment; filename="Plausible.zip"|)
+      |> send_chunked(200)
+
+    DBConnection.run(ch(Plausible.ClickhouseRepo), fn ch_conn ->
+      ch_conn
+      |> Plausible.Exports.stream_archive(
+        Plausible.Exports.export_queries(site.id, extname: ".csv"),
+        format: "CSVWithNames"
```
Comment (ruslandoga) [inline lib/plausible_web/controllers/site_controller.ex:879]: Trying out different combinations of compression and format on `(2023-01-01 -- now) plausible.io` data I got the following results

```elixir
  # .<format>.<compression> <time took>, <zip size>
  # .csv 4145ms, 33.5MB
  # .csv.zst 6349ms, 7.1MB
  # .csv.gz 5805ms, 7.7MB
  # .csv.lz4, 5721ms, 8.5MB
  # .ch, 6076ms, 30.4MB
  # .ch.zst, 5434ms, 5.2MB
  # .ch.gz, 4415ms, 5.8MB
  # .ch.lz4, 5382ms, 7.3MB

  def data_export(conn, _params) do
    site = conn.assigns.site

    conn =
      conn
      |> put_resp_content_type("application/zip")
      |> put_resp_header("content-disposition", ~s|attachment; filename="Plausible.zip"|)
      |> send_chunked(200)

    started_at = System.monotonic_time(:millisecond)

    DBConnection.run(ch(Plausible.ClickhouseRepo), fn ch_conn ->
      ch_conn
      |> Plausible.Exports.stream_archive(
        Plausible.Exports.export_queries(site.id, extname: ".ch.lz4"),
        format: "Native",
        headers: [{"accept-encoding", "lz4"}],
        settings: [enable_http_compression: 1]
      )
      |> Stream.each(fn data -> chunk(conn, data) end)
      |> Stream.run()
    end)

    IO.inspect(System.monotonic_time(:millisecond) - started_at, label: "exported in ms")

    conn
  end
```

File: lib/plausible/exports.ex [hunk 4]
```diff
@@ -0,0 +1,305 @@
...
+      selected_as(fragment("toDate(?)", unquote(timestamp)), :date)
+    end
+  end
+
+  defmacrop visit_duration(t) do
+    quote do
+      selected_as(
+        fragment(
+          "toUInt32(round(ifNotFinite(?,0)))",
+          sum(unquote(t).sign * unquote(t).duration) / sum(unquote(t).sign)
```
Comment (ruslandoga) [inline lib/plausible/exports.ex:41]: when does `sum(sign)` become 0?

Comment (ruslandoga) [reply lib/plausible/exports.ex:41]:
```sql
SELECT *
FROM
(
    SELECT
        session_id,
        sum(sign) AS sign,
        count() AS count
    FROM sessions_v2
    GROUP BY session_id
)
WHERE sign = 0
```
shows that all such sessions have two events, with the second one cancelling the first.

File: lib/plausible/exports.ex [hunk 5]
```diff
@@ -0,0 +1,305 @@
...
+
+  defmacrop visits(t) do
+    quote do
+      selected_as(sum(unquote(t).sign), :visits)
+    end
+  end
+
+  defmacrop bounces(t) do
+    quote do
+      selected_as(sum(unquote(t).sign * unquote(t).is_bounce), :bounces)
```
Comment (ruslandoga) [inline lib/plausible/exports.ex:62]: This sometimes evaluates to -1.

Comment (ruslandoga) [reply lib/plausible/exports.ex:62]:
```sql
SELECT *
FROM
(
    SELECT
        session_id,
        sum(sign * is_bounce) AS bounces
    FROM sessions_v2
    WHERE site_id = 37
    GROUP BY session_id
)
WHERE bounces < 0
```
shows 52 such sessions.

File: lib/plausible_web/live/plugins/s3_import.ex [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+    if entry.done? do
+      # consume_uploaded_entry(socket, entry, fn _meta -> {:ok, nil} end)
+
+      Oban.insert!(
+        Plausible.Imports.S3.new(%{
+          "site_id" => socket.assigns.site_id,
+          "email_to" => socket.assigns.email_to,
+          "s3_key" => s3_key(entry, socket)
+        })
+      )
```
[the remark this answers is not in the document]
Comment (ruslandoga) [reply lib/plausible_web/live/plugins/s3_import.ex]: https://github.com/plausible/analytics/pull/3692/commits/0100d2cc2021b858702ceb404e14feedf8cc4517

File: lib/plausible/imports/s3.ex [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+      table = Path.rootname(filename)
+
+      # TODO
+      s3_url = todo_fix_s3_url_for_docker(s3_url)
+
+      statement =
+        """
+        INSERT INTO {table:Identifier} \
+        SELECT {site_id:UInt64} AS site_id, * \
+        FROM s3({s3_url:String},{s3_access_key_id:String},{s3_secret_access_key:String},{s3_format:String},{s3_structure:String})\
```
Comment (ruslandoga) [inline lib/plausible/imports/s3.ex:38]: We can also use `url` table function with a signed url from `Plausible.S3.signed_url/1`

File: lib/plausible/imports/s3.ex [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+
+    :ok
+  end
+
+  # TODO
+  defp todo_fix_s3_url_for_docker(s3_url) do
+    s3_url
+    |> URI.parse()
+    |> Map.update!(:host, fn
+      "localhost" -> "172.19.0.2"
```
Comment (ruslandoga) [inline lib/plausible/imports/s3.ex:83]: This makes MinIO container visible in the ClickHouse container.

File: config/runtime.exs [hunk 9]
```diff
@@ -525,8 +525,7 @@ base_queues = [
   domain_change_transition: 1,
   check_accept_traffic_until: 1,
   s3_data_export: 1,
-  s3_data_import: 1,
-  s3_cleaner: 10
```
Comment (ruslandoga) [inline config/runtime.exs]: - [ ] bucket lifecycle can be used instead

File: config/.env.dev [hunk 10]
```diff
@@ -20,3 +20,12 @@ GOOGLE_CLIENT_SECRET=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
...
 SITE_DEFAULT_INGEST_THRESHOLD=1000000
+
+S3_DISABLED=false
+S3_ACCESS_KEY_ID=minioadmin
+S3_SECRET_ACCESS_KEY=minioadmin
+S3_REGION=us-east-1
+S3_ENDPOINT=http://localhost:9000
+S3_EXPORTS_BUCKET=exports
+S3_IMPORTS_BUCKET=imports
+S3_HOST_FOR_CLICKHOUSE=172.19.0.4
```
Comment (ruslandoga) [inline config/.env.dev:31]: - [ ] make it `plausible_s3` or something

File: lib/plausible/s3/client.ex [hunk 11]
```diff
@@ -0,0 +1,17 @@
+defmodule Plausible.S3.Client do
+  @moduledoc false
+  @behaviour ExAws.Request.HttpClient
+
+  @impl true
+  def request(method, url, body, headers, opts) do
+    req = Finch.build(method, url, headers, body)
+
+    case Finch.request(req, Plausible.Finch, opts) do
```
Comment (ruslandoga) [inline lib/plausible/s3/client.ex:9]: Should it be `Plausible.HTTPClient.request`? And if so, should `Plausible.Sentry.Client` be updated or are they different (Sentry is more meta, whereas S3 ops are part of the normal application flow).

File: lib/plausible/stats/imported.ex [hunk 12]
```diff
@@ -453,8 +461,8 @@ defmodule Plausible.Stats.Imported do
 
   defp select_joined_metrics(q, [:views_per_visit | rest]) do
     q
-    |> select_merge([s, _i], %{
-      views_per_visit: s.views_per_visit
+    |> select_merge([s, i], %{
+      views_per_visit: s.views_per_visit + i.views_per_visit
```
Comment (ruslandoga) [inline lib/plausible/stats/imported.ex:465]: needs to be weighted

File: assets/js/dashboard/stats/graph/top-stats.js [hunk 13]
```diff
@@ -83,8 +83,7 @@ export default class TopStats extends React.Component {
           {this.topStatNumberLong(stat.name, stat.value)} {statName}
         </div>}
 
-        {stat.name === 'Current visitors' && <p className="font-normal text-xs">Last updated <SecondsSinceLastLoad lastLoadTimestamp={lastLoadTimestamp}/>s ago</p>}
-        {stat.name === 'Views per visit' && showingImported && <p className="font-normal text-xs whitespace-nowrap">Based only on native data</p>}
+        {stat.name === 'Current visitors' && <p className="font-normal text-xs">Last updated <SecondsSinceLastLoad lastLoadTimestamp={lastLoadTimestamp} />s ago</p>}
```
Comment (ruslandoga) [inline assets/js/dashboard/stats/graph/top-stats.js:86]: I'll move these changes to https://github.com/plausible/analytics/pull/3792

File: lib/plausible/imported/csv_importer.ex [hunk 14]
```diff
@@ -16,10 +16,116 @@ defmodule Plausible.Imported.CSVImporter do
...
+      DBConnection.run(pool, fn conn ->
+        File.stream!("imported_browsers.csv", _64kb = 64000)
+        |> Stream.into(import_stream(conn, _site_id = 12, _table = "imported_browsers", _format = "CSVWithNames"))
+        |> Stream.run()
+      end)
+
+  """
+  @spec import_stream(DBConnection.t(), pos_integer, String.t(), String.t(), [Ch.query_option()]) ::
+          Ch.Stream.t()
+  def import_stream(conn, site_id, table, format, opts \\ []) do
```
Comment (ruslandoga) [inline lib/plausible/imported/csv_importer.ex:81]: this one is used in self-hosted only. TODO

**Comment by ruslandoga** [comment]
I'm going to split this PR, one for CSVImporter, one for S3 Exports, one for direct exports, one for direct imports.

**Comment by ruslandoga** [comment]
👋 @johanrosenson 

No, the data would be pre-aggregated and use the same tables that are used for the current Google Analytics imports.

**Comment by purcell** [comment]
So they wouldn't have sufficient detail to switch seamlessly between the self-hosted and SaaS versions, or vice versa?

**Comment by ruslandoga** [comment]
👋 @purcell 

I think it would be similar to switching from Google Analytics to Plausible. Some details do get lost, like hourly charts. And due to how Plausible is calculating session data, visitors count get over-estimated in exports. Here're some screenshots of my WIP for plausible.io data after export/import flow.

Last year:

Screenshot 2024-02-26 at 19 12 04
Screenshot 2024-02-26 at 19 11 48

December 2023:

Screenshot 2024-02-26 at 19 19 25
Screenshot 2024-02-26 at 19 19 27

December 27, 2023

Screenshot 2024-02-26 at 19 19 39
Screenshot 2024-02-26 at 19 19 42

**Pull request closed without merging**